### PR TITLE
Changing letsencrypt module to use ACME v2 protocol (update)

### PR DESCRIPTION
**Pull request by felixfontein**
##### SUMMARY
This is a continuation of #34541, with all the missing things.

##### ISSUE TYPE
Feature/Bugfix Pull Request

##### COMPONENT NAME
letsencrypt

##### ANSIBLE VERSION
```
2.5, devel
```

### CODE REVIEW

**Comment by felixfontein** [comment]
First step: adding a warning.

**Comment by felixfontein** [comment]
@resmo: I think we should leave this one open to collect possibly necessary updates/bugfixes/... until Feb 7.

**Comment by resmo** [comment]
ack

**Comment by resmo** [comment]
@felixfontein don't forget the
~~~diff
-            acme_version=dict(required=False, default=1, type='int'),
+            acme_version=dict(required=False, default=1, choices=[1, 2], type='int'),
~~~

:)

**Comment by felixfontein** [comment]
Thanks for reminding me! Just added that.

**Comment by felixfontein** [comment]
The Let's Encrypt v2 staging endpoint now sends certificate chains (letsencrypt/boulder#3291); I've tested that (using #34328) and it works fine. \o/

**Comment by felixfontein** [comment]
I've also removed the `errors` field support again: it won't be implemented in Boulder, and has been removed from the specs; see letsencrypt/boulder#3339.

File: lib/ansible/modules/web_infrastructure/letsencrypt.py [hunk 1]
```diff
@@ -238,6 +238,11 @@
       returned: changed
       type: string
       sample: IlirfxKKXA...17Dt3juxGJ-PCt92wr-oA
+    record:
+      description: the full DNS record's name for the challenge
+      returned: changed and challenge is dns-01
+      type: string
+      sample: _acme-challenge.example.com
```
Comment (resmo) [inline lib/ansible/modules/web_infrastructure/letsencrypt.py:285]: please add `version_added: 2.5`

**Comment by felixfontein** [comment]
I added code which for DNS challenges also inserts a `record` field, which contains the rull record name to create (excluding trailing dots, which might be required for some DNS modules). This is to simplify obtaining wildcard certs:

If you want to get a certificate for `*.example.com` and `example.com` at the same time, you need to add *two* TXT records for `_acme-challenge.example.com`: one is the challenge for `*.example.com`, and one the challenge for `example.com`. (This collision only happens in precisely this combination of wildcards and non-wildcards.) The problem is that some Ansible DNS modules, like [route53](https://docs.ansible.com/ansible/latest/route53_module.html), require the user to specify all values for one record at the same time.

I've tried writing some generic Ansible code (for my `letsencrypt` role) which collects the challenges per zone, and creates one (combined) entry for them; the code is pretty horrible to read, but works:
```
  - debug:
      msg: "record={{ result.challenge_data.items() | selectattr(0, 'match', '^(?:\\*\\.|)' ~ (item | regex_escape()) ~ '$') | map(attribute='1.dns-01.resource') | first }}.{{ item }}
            values={{ result.challenge_data.items() | selectattr(0, 'match', '^(?:\\*\\.|)' ~ (item | regex_escape()) ~ '$') | map(attribute='1.dns-01.resource_value') | list }}"
    with_items: "{{ result.challenge_data.keys() | map('regex_replace', '^\\*\\.', '') | unique }}"
```
With this change, this can be heavily simplified:
```
  - debug:
      msg: "record={{ item }}
            values={{ result.challenge_data.items() | selectattr('1.dns-01.record', 'equalto', item) | map(attribute='1.dns-01.resource_value') | list }}"
    with_items: "{{ result.challenge_data.values() | map(attribute='dns-01.record') | unique }}"
```
I think this simplification is worth adding this new field.

File: lib/ansible/modules/web_infrastructure/letsencrypt.py (hunk 1 above)
Comment (felixfontein) [reply lib/ansible/modules/web_infrastructure/letsencrypt.py:285]: Ah, yes, sorry, forgot that :)

**Comment by felixfontein** [comment]
Even nicer would be a completely new return field, like `challenge_data_dns`, which already groups these values together so it would reduce to something like
```
  - debug:
      msg: "record={{ item.key }}
            values={{ item.value }}"
    with_dict: "{{ result.challenge_data_dns }}"
```
Any opinions?

**Comment by felixfontein** [comment]
@resmo: I added 123e146bb9d719c0304b1d3c57bd0d0556d4d495 to allow disabling of certificate validation when talking to the ACME server. This should only ever be needed for testing (using a local instance of the [Pebble server](https://github.com/letsencrypt/Pebble)). For now I also documented that setting in the metadata, but I'm not sure whether that's a good idea (I'm afraid people actually disable validation because they try random config settings when something doesn't work as expected). Is it acceptable if I add this parameter to `AnsibleModule`, but don't mention it in the `DOCUMENTATION` metadata?

**Comment by resmo** [comment]
@felixfontein I see your concerns, but we should keep it in docs (probably will the docs gets generated of these params in later ansible versions... so we can not hide it anyway). And I don't like to dictate "security", the user, but one thing is to put a big fat warning to the console. 

E.g. 
~~~python
if not module.params.get('validate_certs'):
    module.warn(warning="some warnings here")
~~~

I even thought it would make sense to put this directly in the fetch_url in utils...

**Comment by felixfontein** [comment]
Ah, that's an excellent idea! I'll do that right away.

**Comment by felixfontein** [comment]
(And yes, having that in `fetch_url` would be even better...)

**Comment by felixfontein** [comment]
(In case #35399 is merged for Ansible 2.5, we should probably remove the `module.warn()` call, but not the more explicit warning in the documentation.)

**Comment by resmo** [comment]
@felixfontein anything left on your list for this PR?

**Comment by felixfontein** [comment]
@resmo I don't think so; except more testing when new versions of the staging endpoint are deployed, of course ;) but I think we can merge now, I'll create a new PR if something shows up during testing...

**Review by resmo — APPROVED** [review]
shipit

**Comment by felixfontein** [comment]
Thanks for reviewing and merging!

**Pull request merged**